### PR TITLE
WT-12971 Fix assertion failure: 'onpage_upd != ((void*)0) || supd_restore' on FLCS eviction

### PR DESCRIPTION
**Pull request by pmacko86**
This PR removes adding a tombstone to empty, implicitly created FLCS records, because doing so causes problems during reconciliation (please refer to the ticket and/or the new comment in the code for details).

### CODE REVIEW

**Review by sueloverso — APPROVED** [review]
LGTM although I'm not deeply knowledgeable about FLCS.

File: src/btree/bt_cursor.c [hunk 1]
```diff
@@ -1478,14 +1478,20 @@ __wt_btcur_remove(WT_CURSOR_BTREE *cbt, bool positioned)
...
+             * We don't want to add a tombstone to an empty record, because doing so could create a
+             * problem during reconciliation: If this is a part of a prepared transaction, we would
+             * need to save the update so that we can then put it into the history store, but in
+             * this case, there is nothing to save.
              *
              * Correct the btree cursor's location: the search will have pointed us at the
              * previous/next item, and that's not correct.
              */
             cbt->recno = cursor->recno;
-            ret = __cursor_col_modify(cbt, NULL, WT_UPDATE_TOMBSTONE);
```
Comment (luke-pearson) [inline src/btree/bt_cursor.c:1488]: Changing this feels like it's moving away from a core design decision in FLCS, I'm not super familiar with why it was done this way. We may want to ask around for an FLCS expert to review. I am curious, why fix it here and not in reconciliation?

Comment (pmacko86) [reply src/btree/bt_cursor.c:1488]: I don't think this breaks anything in FLCS, as the implicit records are already created at this point, but let's still ask.

I spent some time trying to figure out how to fix it in reconciliation, and couldn't figure it out. If I took care of one place where it broke reconciliation, it caused problems in another place, which tells me it would have to be a bigger fix.

Comment (pmacko86) [reply src/btree/bt_cursor.c:1488]: Yes, that's also how I understand this, but let me rephrase it my words to see if we are on the same page: As the record is implicitly created and has nothing in it, delete by itself has no effect—reading the value of the record before delete returns 0, and reading the value after delete also returns 0. And the record still exists after the removal, because it has been implicitly created. And calling remove on an implicitly created key returns 0 (as opposed to WT_NOTFOUND), because the record implicitly exists.

File: src/btree/bt_cursor.c [hunk 2]
```diff
@@ -1478,14 +1478,20 @@ __wt_btcur_remove(WT_CURSOR_BTREE *cbt, bool positioned)
             ret = __cursor_col_modify(cbt, NULL, WT_UPDATE_TOMBSTONE);
         } else if (__cursor_fix_implicit(btree, cbt)) {
             /*
-             * Creating a record past the end of the tree in a fixed-length column-store implicitly
-             * fills the gap with empty records, delete the record.
+             * We are deleting an implicitly created record (which exists because creating a record
+             * past the end of the tree in a fixed-length column-store implicitly fills the gap with
+             * empty records). But because this record is empty, deleting it is a no-op. But keep
```
Comment (luke-pearson) [inline src/btree/bt_cursor.c:1483]: I would rephrase this comment a bit to make it more detailed.

"We are deleting an implicitly created record, which exists as another record was added further along and as such FLCS fills the gap with implicit records. Deleting implicit records doesn't make sense, and as such this is a no-op. An implicit record returns 0 when searched for. If we were to add a tombstone here it would then mean the search returns WT_NOTFOUND, which is then translated back into 0. So if we add a tombstone we return 0, if we don't also 0. As such we choose to not add a tombstone here.

Not adding a tombstone, which is what this code did previously, fixes reconciliation behaviour with regards to prepared transactions. Prepared transactions must save updates when reconciling the update chain, but a single tombstone on the chain cannot be saved, this causes an assertion to fire."

Feel free to pick parts of that comment to apply. I think your phrasing is kinda too repetitive and not explicit enough. Also the phrase "but keep the return value as 0" assumes a lot of knowledge.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/btree/bt_cursor.c (hunk 2 above)
Comment (pmacko86) [reply src/btree/bt_cursor.c]: Thanks for the suggestion! I rephrased this a bit and updated the comment accordingly in [2f711dc](https://github.com/wiredtiger/wiredtiger/pull/10787/commits/2f711dc726e85e1f226db2f5eaad5bbed1020a7b).

**Review by luke-pearson — APPROVED** [review]
LGTM.
